R/misc.R

```diff
@@ -0,0 +1,73 @@
+#' Helper function to print out log messages into Shiny
+#' using cat() and stderr(), as described on
+#' https://shiny.rstudio.com/articles/debugging.html
+#'
+#' Because this can print a lot, it's silent unless
+#' the shiny.router.debug option is set.
+#'
+#' @param ... All params get passed through to cat().
+#' They're automatically wrapped in shiny::isolate(),
+#' so you can print reactive values here without too
+#' much worry.
+log_msg <- function(...) {
+  if (getOption("shiny.router.debug", default = FALSE)) {
+    shiny::isolate(
+      do.call(
+        "cat",
+        alist(
+          file = stderr(),
+          ...
+        )
+      )
+    )
+    cat("\n")
+  }
+}
+
+
+#' Formats a URL fragment into a hashpath starting with "#!/"
+#'
+#' @param hashpath character with hash path
+#'
+#' @example
+#' cleanup_hashpath("/abc") #  "#!/abc"
+cleanup_hashpath <- function(hashpath) {
+  hashpath = hashpath[1]
+
+  # Already correctly formatted.
+  if (substr(hashpath, 1, 3) == "#!/") {
+    return(hashpath)
+  }
+
+  # We remove any partial hashbang path morker from the start
+  # of the string, then add back the full one.
+  slicefrom <- 1
+  if (substr(hashpath, slicefrom, slicefrom) == "#") {
+    slicefrom <- slicefrom + 1
+  }
+  if (substr(hashpath, slicefrom, slicefrom) == "!") {
+    slicefrom <- slicefrom + 1
+  }
+  if (substr(hashpath, slicefrom, slicefrom) == "/") {
+    slicefrom <- slicefrom + 1
+  }
```

Comment:
I imagine there are a lot of corner cases resistant for these 3 if statements. E.g. `##!//url` is still valid part of the url for the browser. How about something like this:

```
slicefrom <- 1
while (substr(hashpath, slicefrom, slicefrom) %in% c("#", "!", "/")) slicefrom <- slicefrom + 1
```

What do you think? Also you have one typo in the comment above ("hashbang path morker" <-- "o")

Reply:
Thanks, that's useful!